ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -241,8 +315,351 @@ public void testValidDataSizeAfterDeletes() throws InterruptedException, StoreEx
+    blobStoreStats.close();
+  }
+
+  @Test
+  public void testBucketingWithLowIndexEntries() throws InterruptedException, StoreException, IOException {
+    state.destroy();
+    assertTrue(tempDir.getAbsolutePath() + " could not be deleted", StoreTestUtils.cleanDirectory(tempDir, true));
+    tempDir = StoreTestUtils.createTempDirectory("blobStoreStatsDir-" + UtilsTest.getRandomString(10));
+    state = new CuratedLogIndexState(isLogSegmented, tempDir, false, false);
+    int bucketCount = bucketingEnabled ? 1 : 0;
+    BlobStoreStats blobStoreStats = setupBlobStoreStats(bucketCount, 0);
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0));
+    state.addPutEntries(3, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time);
+    blobStoreStats = setupBlobStoreStats(bucketCount, 0);
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0));
+  }
+
+  /**
+   * Test to verify that separate scan is triggered to answer the request after waiting for scan timed out.
+   * @throws InterruptedException
+   * @throws StoreException
+   */
+  @Test(timeout = 1000)
+  public void testBucketingWaitTimeout() throws InterruptedException, StoreException {
+    assumeTrue(bucketingEnabled);
+    CountDownLatch scanStartedLatch = new CountDownLatch(1);
+    CountDownLatch scanHoldLatch = new CountDownLatch(1);
+    MockThrottler mockThrottler = new MockThrottler(scanStartedLatch, scanHoldLatch);
+    throttlers.put(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, mockThrottler);
+    BlobStoreStats blobStoreStats =
+        new BlobStoreStats("", state.index, 10, BUCKET_SPAN_IN_MS, 0, QUEUE_PROCESSOR_PERIOD_IN_SECS, 0, state.time,
+            indexScannerScheduler, queueProcessorScheduler, diskIOScheduler, METRICS);
+    // proceed only when the scan is started
+    assertTrue("IndexScanner took too long to start", scanStartedLatch.await(5, TimeUnit.SECONDS));
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0));
+    advanceTimeToNextSecond();
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0));
+    scanHoldLatch.countDown();
+    blobStoreStats.close();
+  }
+
+  /**
+   * Test to verify that once the {@link BlobStoreStats} is closed (or closing), requests throw {@link StoreException}.
+   * @throws InterruptedException
+   */
+  @Test(timeout = 1000)
+  public void testRequestOnClosing() throws InterruptedException {
+    int bucketCount = bucketingEnabled ? 1 : 0;
+    BlobStoreStats blobStoreStats = setupBlobStoreStats(bucketCount, 0);
+    blobStoreStats.close();
+    try {
```

Comment:
Can you cover the scenario when close is called when index scanning is in progress(case 1: before first checkpoint case2: after 1st, before 2nd checkpoint. case3: when request range was outside the bucket forcast and an on-demand scan was triggered)